File: src/main/java/dev/tricked/hardermc/tools/CustomRecipeManager.java
package dev.tricked.hardermc.tools;

import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerPickupItemEvent;
import org.bukkit.inventory.Recipe;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import static org.bukkit.Bukkit.getServer;

public class CustomRecipeManager implements Listener {
    private List<CustomRecipe> customRecipes = new ArrayList<>();

    public void addCustomRecipe(CustomRecipe data) {
        Bukkit.removeRecipe(data.key);
        getServer().addRecipe(data.recipe);
        customRecipes.add(data);
    }


    @EventHandler(priority = EventPriority.MONITOR)
    public void onPlayerPickupItem(PlayerPickupItemEvent event) {
        var type = event.getItem().getItemStack().getType();
        for (CustomRecipe recipe : customRecipes) {
            if (recipe.triggerIngredients.contains(type)) {
                event.getPlayer().discoverRecipe(recipe.key);
            }
        }
    }
}

File: src/main/kotlin/dev/tricked/hardermc/features/CustomRecipeManager.kt
/*
 * Copyright (c) Tricked-dev 2024.
 */

package dev.tricked.hardermc.features

import dev.tricked.hardermc.HarderMC
import dev.tricked.hardermc.utilities.*
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.entity.EntityPickupItemEvent

@Name("Custom Recipe Manager")
@Description("Enables the custom recipes and the obtaining of the recipes")
class CustomRecipeManager(plugin: HarderMC) : BaseTool(plugin) {
    private val customRecipes: MutableList<CustomRecipe> = ArrayList()

    val customArmorEnabled: Boolean by ConfigProperty(plugin, configPrefix, true)
    val customEnchantmentTableEnabled: Boolean by ConfigProperty(plugin, configPrefix, true)
    val customShieldEnabled: Boolean by ConfigProperty(plugin, configPrefix, true)
    val helpBookEnabled: Boolean by ConfigProperty(plugin, configPrefix, true)
    val customEyeOfEnderEnabled: Boolean by ConfigProperty(plugin, configPrefix, true)

    private fun addCustomRecipe(data: CustomRecipe) {
        Bukkit.removeRecipe(data.key!!)
        Bukkit.getServer().addRecipe(data.recipe)
        customRecipes.add(data)
    }

    fun addCustomRecipes(vararg data: CustomRecipe) {
        if (!enabled) return

        for (recipe in data) {
            getLog().info("Adding custom recipe: " + recipe.key)
            addCustomRecipe(recipe)
        }
    }

    @EventHandler(priority = EventPriority.MONITOR)
    fun onPlayerPickupItem(event: EntityPickupItemEvent) {
        if (!enabled) return

        val type = event.item.itemStack.type
        for (recipe in customRecipes) {
            if (recipe.triggerIngredients.contains(type)) {
                if (event.entity is Player) {
                    (event.entity as Player).discoverRecipe(recipe.key!!)
                }
            }
        }
    }
}
